File: xauth.go
package main

import (
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/go-resty/resty/v2"
	"github.com/tidwall/gjson"
)

const (
	twitterAuthority = "twitter.com"
	twitterOrigin    = "https://twitter.com"
	twitterAPIBase   = "https://twitter.com/i/api/2"
	userAgent        = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
	authorization    = "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA"
	maxRetries       = 3
	retryInterval    = 1 * time.Second
)

// AccountState 定义账户状态映射
var accountState = map[int64]string{
	32:  "Bad Token",
	64:  "SUSPENDED",
	141: "SUSPENDED",
	326: "LOCKED",
}

type XAuth struct {
	client    *resty.Client
	client2   *resty.Client
	authToken string
}

// NewTwitter 创建新的XAuth实例
func NewTwitter(authToken string) *XAuth {
	baseHeaders := map[string]string{
		"user-agent": userAgent,
	}

	// 创建主客户端
	client := resty.New().
		SetHeaders(baseHeaders).
		SetHeaders(map[string]string{
			"authority":             twitterAuthority,
			"origin":                twitterOrigin,
			"x-twitter-auth-type":   "OAuth2Session",
			"x-twitter-active-user": "yes",
			"authorization":         authorization,
		}).
		SetRetryCount(maxRetries).
		SetRetryWaitTime(retryInterval).
		SetTimeout(10 * time.Second).
		SetCookie(&http.Cookie{Name: "auth_token", Value: authToken})

	// 创建次要客户端
	client2 := resty.New().
		SetHeaders(baseHeaders).
		SetRetryCount(maxRetries).
		SetRetryWaitTime(retryInterval).
		SetTimeout(10 * time.Second).
		SetCookie(&http.Cookie{Name: "auth_token", Value: authToken})

	return &XAuth{
		client:    client,
		client2:   client2,
		authToken: authToken,
	}
}

func (x *XAuth) get_twitter_token(oauth_token string) (string, error) {
	if oauth_token == "" {
		return "", fmt.Errorf("oauth_token不能为空")
	}

	params := map[string]string{
		"oauth_token": oauth_token,
	}

	resp, err := x.client2.R().SetQueryParams(params).
		Get("https://api.x.com/oauth/authenticate")
	if err != nil {
		return "", fmt.Errorf("请求Twitter认证失败: %w", err)
	}

	if !strings.Contains(resp.String(), "authenticity_token") {
		if strings.Contains(resp.String(), "The request token for this page is invalid") {
			return "", fmt.Errorf("请求oauth_token无效")
		}
		return "", fmt.Errorf("响应中未找到authenticity_token")
	}

	parts := strings.Split(resp.String(), "name=\"authenticity_token\" value=\"")
	if len(parts) < 2 {
		parts = strings.Split(resp.String(), "name=\"authenticity_token\" type=\"hidden\" value=\"")
		if len(parts) < 2 {
			return "", fmt.Errorf("authenticity_token格式错误")
		}
	}

	token := strings.Split(parts[1], "\"")[0]
	if token == "" {
		return "", fmt.Errorf("获取到的authenticity_token为空")
	}

	return token, nil
}

func (x *XAuth) Oauth1(oauth_token string) (string, error) {
	authenticityToken, err := x.get_twitter_token(oauth_token)
	if err != nil {
		return "", fmt.Errorf("获取Twitter token失败: %w", err)
	}

	data := map[string]string{
		"authenticity_token": authenticityToken,
		"oauth_token":        oauth_token,
	}

	resp, err := x.client2.R().SetFormData(data).
		Post("https://x.com/oauth/authorize")
	if err != nil {
		return "", fmt.Errorf("OAuth1认证请求失败: %w", err)
	}

	if !strings.Contains(resp.String(), "oauth_verifier") {
		if strings.Contains(resp.String(), "This account is suspended.") {
			return "", fmt.Errorf("该账户已被封禁")
		}
	}

	parts := strings.Split(resp.String(), "oauth_verifier=")
	if len(parts) < 2 {
		return "", fmt.Errorf("oauth_verifier格式错误")
	}

	verifier := strings.Split(parts[1], "\"")[0]
	if verifier == "" {
		return "", fmt.Errorf("获取到的oauth_verifier为空")
	}

	return verifier, nil
}

func (x *XAuth) get_auth_code(params map[string]string) (string, error) {
	if len(params) == 0 {
		return "", fmt.Errorf("参数不能为空")
	}

	resp, err := x.client.R().
		SetQueryParams(params).
		Get(twitterAPIBase + "/oauth2/authorize")
	if err != nil {
		return "", fmt.Errorf("请求auth_code失败: %w", err)
	}

	// 处理请求限制
	if resp.StatusCode() == http.StatusTooManyRequests {
		time.Sleep(retryInterval)
		return x.get_auth_code(params)
	}

	// 处理CSRF token
	code := gjson.Get(resp.String(), "code")
	if code.Exists() && code.Int() == 353 {
		for _, cookie := range resp.Cookies() {
			if cookie.Name == "ct0" {
				x.client.SetHeader("x-csrf-token", cookie.Value)
				return x.get_auth_code(params)
			}
		}
		return "", fmt.Errorf("未找到ct0 cookie")
	}

	// 检查账户状态
	if errCode := gjson.Get(resp.String(), "errors.0.code"); errCode.Exists() {
		if state, exists := accountState[errCode.Int()]; exists {
			return "", fmt.Errorf("token状态错误: %s", state)
		}
	}

	// 获取auth_code
	authCode := gjson.Get(resp.String(), "auth_code")
	if !authCode.Exists() {
		return "", fmt.Errorf("响应中未找到auth_code")
	}

	return authCode.String(), nil
}

func (x *XAuth) Oauth2(params map[string]string) (string, error) {
	authCode, err := x.get_auth_code(params)
	if err != nil {
		return "", fmt.Errorf("获取auth_code失败: %w", err)
	}

	data := map[string]string{
		"approval": "true",
		"code":     authCode,
	}

	resp, err := x.client.R().SetFormData(data).
		Post(twitterAPIBase + "/oauth2/authorize")
	if err != nil {
		return "", fmt.Errorf("OAuth2认证请求失败: %w", err)
	}

	// 处理请求限制
	if resp.StatusCode() == http.StatusTooManyRequests {
		time.Sleep(retryInterval)
		return x.Oauth2(params)
	}

	if !strings.Contains(resp.String(), "redirect_uri") {
		return "", fmt.Errorf("响应中未找到redirect_uri")
	}

	return authCode, nil
}

File: java/src/main/java/com/xauth/XAuth.java
package com.xauth;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.NameValuePair;
import org.apache.http.client.CookieStore;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpRequestBase;
import org.apache.http.client.protocol.HttpClientContext;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.cookie.Cookie;
import org.apache.http.impl.client.BasicCookieStore;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.cookie.BasicClientCookie;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Twitter OAuth认证工具类
 * 支持OAuth1和OAuth2认证流程
 */
public class XAuth {
    private static final Logger logger = LoggerFactory.getLogger(XAuth.class);

    // 常量定义
    private static final String TWITTER_AUTHORITY = "twitter.com";
    private static final String TWITTER_ORIGIN = "https://twitter.com";
    private static final String TWITTER_API_BASE = "https://twitter.com/i/api/2";
    private static final String USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36";
    private static final String AUTHORIZATION = "Bearer AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA";
    private static final int MAX_RETRIES = 3;
    private static final int RETRY_INTERVAL = 1000; // milliseconds

    // 账户状态码映射
    private static final Map<Integer, String> ACCOUNT_STATE = new HashMap<>();
    static {
        ACCOUNT_STATE.put(32, "Bad Token");
        ACCOUNT_STATE.put(64, "SUSPENDED");
        ACCOUNT_STATE.put(141, "SUSPENDED");
        ACCOUNT_STATE.put(326, "LOCKED");
    }

    private final String authToken;
    private final CloseableHttpClient client;
    private final CloseableHttpClient client2;
    private final CookieStore cookieStore;
    private final HttpClientContext context;
    private final ObjectMapper objectMapper;

    /**
     * 初始化XAuth实例
     *
     * @param authToken Twitter认证token
     * @throws IllegalArgumentException 当auth_token为空时
     */
    public XAuth(String authToken) {
        if (authToken == null || authToken.isEmpty()) {
            throw new IllegalArgumentException("auth_token不能为空");
        }

        this.authToken = authToken;
        this.objectMapper = new ObjectMapper();
        this.cookieStore = new BasicCookieStore();

        // 设置auth_token cookie
        BasicClientCookie cookie = new BasicClientCookie("auth_token", authToken);
        cookie.setDomain(TWITTER_AUTHORITY);
        cookie.setPath("/");
        cookieStore.addCookie(cookie);

        // 创建HTTP上下文
        this.context = HttpClientContext.create();
        context.setCookieStore(cookieStore);

        // 创建HTTP客户端
        RequestConfig requestConfig = RequestConfig.custom()
                .setConnectTimeout(10000)
                .setSocketTimeout(10000)
                .build();

        this.client = HttpClients.custom()
                .setDefaultCookieStore(cookieStore)
                .setDefaultRequestConfig(requestConfig)
                .build();

        this.client2 = HttpClients.custom()
                .setDefaultCookieStore(cookieStore)
                .setDefaultRequestConfig(requestConfig)
                .build();
    }

    /**
     * 设置请求头
     *
     * @param request HTTP请求
     * @param includeTwitterHeaders 是否包含Twitter特定的headers
     */
    private void setHeaders(HttpRequestBase request, boolean includeTwitterHeaders) {
        request.setHeader("User-Agent", USER_AGENT);

        if (includeTwitterHeaders) {
            request.setHeader("authority", TWITTER_AUTHORITY);
            request.setHeader("origin", TWITTER_ORIGIN);
            request.setHeader("x-twitter-auth-type", "OAuth2Session");
            request.setHeader("x-twitter-active-user", "yes");
            request.setHeader("authorization", AUTHORIZATION);
        }
    }

    /**
     * 处理响应状态
     *
     * @param response HTTP响应
     * @throws IOException 当请求失败时
     * @throws InterruptedException 当线程被中断时
     */
    private void handleResponse(HttpResponse response) throws IOException, InterruptedException {
        int statusCode = response.getStatusLine().getStatusCode();
        if (statusCode == HttpStatus.SC_TOO_MANY_REQUESTS) {
            Thread.sleep(RETRY_INTERVAL);
        }
    }

    /**
     * 获取Twitter认证token
     *
     * @param oauthToken OAuth token
     * @return 认证token字符串
     * @throws IOException 当请求失败时
     * @throws InterruptedException 当线程被中断时
     * @throws XAuthException 当token无效或响应格式不正确时
     */
    public String getTwitterToken(String oauthToken) throws IOException, InterruptedException, XAuthException {
        if (oauthToken == null || oauthToken.isEmpty()) {
            throw new XAuthException("oauth_token不能为空");
        }

        URI uri;
        try {
            uri = new URIBuilder("https://api.x.com/oauth/authenticate")
                    .addParameter("oauth_token", oauthToken)
                    .build();
        } catch (URISyntaxException e) {
            throw new XAuthException("构建URI失败", e);
        }

        HttpGet request = new HttpGet(uri);
        setHeaders(request, false);

        try (CloseableHttpResponse response = client2.execute(request, context)) {
            handleResponse(response);
            
            HttpEntity entity = response.getEntity();
            String content = EntityUtils.toString(entity, StandardCharsets.UTF_8);

            if (!content.contains("authenticity_token")) {
                if (content.contains("The request token for this page is invalid")) {
                    throw new XAuthException("请求oauth_token无效");
                }
                throw new XAuthException("响应中未找到authenticity_token");
            }

            // 尝试两种可能的token格式
            String[] tokenMarkers = {
                    "name=\"authenticity_token\" value=\"",
                    "name=\"authenticity_token\" type=\"hidden\" value=\""
            };

            String token = null;
            for (String marker : tokenMarkers) {
                if (content.contains(marker)) {
                    int startIndex = content.indexOf(marker) + marker.length();
                    int endIndex = content.indexOf("\"", startIndex);
                    if (startIndex >= 0 && endIndex > startIndex) {
                        token = content.substring(startIndex, endIndex);
                        break;
                    }
                }
            }

            if (token == null || token.isEmpty()) {
                throw new XAuthException("获取到的authenticity_token为空");
            }

            return token;
        }
    }

    /**
     * 执行OAuth1认证流程
     *
     * @param oauthToken OAuth token
     * @return OAuth验证码
     * @throws IOException 当请求失败时
     * @throws InterruptedException 当线程被中断时
     * @throws XAuthException 当认证失败时
     */
    public String oauth1(String oauthToken) throws IOException, InterruptedException, XAuthException {
        String authenticityToken = getTwitterToken(oauthToken);

        HttpPost request = new HttpPost("https://x.com/oauth/authorize");
        setHeaders(request, false);

        List<NameValuePair> params = new ArrayList<>();
        params.add(new BasicNameValuePair("authenticity_token", authenticityToken));
        params.add(new BasicNameValuePair("oauth_token", oauthToken));
        request.setEntity(new UrlEncodedFormEntity(params, StandardCharsets.UTF_8));

        try (CloseableHttpResponse response = client2.execute(request, context)) {
            handleResponse(response);
            
            HttpEntity entity = response.getEntity();
            String content = EntityUtils.toString(entity, StandardCharsets.UTF_8);

            if (!content.contains("oauth_verifier")) {
                if (content.contains("This account is suspended.")) {
                    throw new XAuthException("该账户已被封禁");
                }
                throw new XAuthException("未找到oauth_verifier");
            }

            Pattern pattern = Pattern.compile("oauth_verifier=([^\"&]+)");
            Matcher matcher = pattern.matcher(content);
            if (matcher.find()) {
                String verifier = matcher.group(1);
                if (verifier == null || verifier.isEmpty()) {
                    throw new XAuthException("获取到的oauth_verifier为空");
                }
                return verifier;
            } else {
                throw new XAuthException("未找到oauth_verifier");
            }
        }
    }

    /**
     * 获取认证码
     *
     * @param params 请求参数
     * @return 认证码
     * @throws IOException 当请求失败时
     * @throws InterruptedException 当线程被中断时
     * @throws XAuthException 当参数无效或响应格式不正确时
     */
    public String getAuthCode(Map<String, String> params) throws IOException, InterruptedException, XAuthException {
        if (params == null || params.isEmpty()) {
            throw new XAuthException("参数不能为空");
        }

        URIBuilder uriBuilder = new URIBuilder();
        uriBuilder.setScheme("https")
                .setHost(TWITTER_AUTHORITY)
                .setPath("/i/api/2/oauth2/authorize");

        for (Map.Entry<String, String> entry : params.entrySet()) {
            uriBuilder.addParameter(entry.getKey(), entry.getValue());
        }

        URI uri;
        try {
            uri = uriBuilder.build();
        } catch (URISyntaxException e) {
            throw new XAuthException("构建URI失败", e);
        }

        HttpGet request = new HttpGet(uri);
        setHeaders(request, true);

        try (CloseableHttpResponse response = client.execute(request, context)) {
            handleResponse(response);
            
            HttpEntity entity = response.getEntity();
            String content = EntityUtils.toString(entity, StandardCharsets.UTF_8);

            JsonNode data;
            try {
                data = objectMapper.readTree(content);
            } catch (Exception e) {
                throw new XAuthException("响应不是有效的JSON格式", e);
            }

            // 处理CSRF token
            if (data.has("code") && data.get("code").asInt() == 353) {
                // 获取ct0 cookie
                Cookie ct0Cookie = null;
                for (Cookie cookie : cookieStore.getCookies()) {
                    if ("ct0".equals(cookie.getName())) {
                        ct0Cookie = cookie;
                        break;
                    }
                }

                if (ct0Cookie != null) {
                    // 重新发起请求，带上CSRF token
                    HttpGet newRequest = new HttpGet(uri);
                    setHeaders(newRequest, true);
                    newRequest.setHeader("x-csrf-token", ct0Cookie.getValue());
                    
                    try (CloseableHttpResponse newResponse = client.execute(newRequest, context)) {
                        handleResponse(newResponse);
                        
                        HttpEntity newEntity = newResponse.getEntity();
                        String newContent = EntityUtils.toString(newEntity, StandardCharsets.UTF_8);
                        
                        try {
                            data = objectMapper.readTree(newContent);
                        } catch (Exception e) {
                            throw new XAuthException("响应不是有效的JSON格式", e);
                        }
                    }
                } else {
                    throw new XAuthException("未找到ct0 cookie");
                }
            }

            // 检查错误
            if (data.has("errors") && data.get("errors").isArray() && data.get("errors").size() > 0) {
                JsonNode error = data.get("errors").get(0);
                if (error.has("code")) {
                    int errorCode = error.get("code").asInt();
                    if (ACCOUNT_STATE.containsKey(errorCode)) {
                        throw new XAuthException("token状态错误: " + ACCOUNT_STATE.get(errorCode));
                    }
                }
            }

            // 获取auth_code
            if (!data.has("auth_code") || data.get("auth_code").isNull()) {
                throw new XAuthException("响应中未找到auth_code");
            }

            return data.get("auth_code").asText();
        }
    }

    /**
     * 执行OAuth2认证流程
     *
     * @param params 请求参数
     * @return 认证码
     * @throws IOException 当请求失败时
     * @throws InterruptedException 当线程被中断时
     * @throws XAuthException 当认证失败时
     */
    public String oauth2(Map<String, String> params) throws IOException, InterruptedException, XAuthException {
        String authCode = getAuthCode(params);

        HttpPost request = new HttpPost(TWITTER_API_BASE + "/oauth2/authorize");
        setHeaders(request, true);
        
        // 获取ct0 cookie并设置CSRF token
        for (Cookie cookie : cookieStore.getCookies()) {
            if ("ct0".equals(cookie.getName())) {
                request.setHeader("x-csrf-token", cookie.getValue());
                break;
            }
        }

        List<NameValuePair> formParams = new ArrayList<>();
        formParams.add(new BasicNameValuePair("approval", "true"));
        formParams.add(new BasicNameValuePair("code", authCode));
        request.setEntity(new UrlEncodedFormEntity(formParams, StandardCharsets.UTF_8));

        try (CloseableHttpResponse response = client.execute(request, context)) {
            handleResponse(response);
            
            HttpEntity entity = response.getEntity();
            String content = EntityUtils.toString(entity, StandardCharsets.UTF_8);

            if (!content.contains("redirect_uri")) {
                throw new XAuthException("响应中未找到redirect_uri");
            }

            return authCode;
        }
    }

    /**
     * 关闭HTTP客户端
     */
    public void close() {
        try {
            if (client != null) {
                client.close();
            }
            if (client2 != null) {
                client2.close();
            }
        } catch (IOException e) {
            logger.error("关闭HTTP客户端失败", e);
        }
    }
} 